scripts/depth/trainer.py

```diff
@@ -0,0 +1,499 @@
+from __future__ import absolute_import, division, print_function
+
+import os
+import sys
+import shutil
+from tqdm import tqdm
+
+import cv2
+import numpy as np
+import time
+
+import json
+
+import mxnet as mx
+from mxnet import gluon, autograd
+import mxnet.numpy as _mx_np
+from mxnet.util import is_np_array
+
+import gluoncv.data.kitti as kitti_dataset
+from gluoncv.data.kitti.kitti_utils import dict_batchify_fn
+from gluoncv.model_zoo import monodepthv2
+from gluoncv.model_zoo.monodepthv2.layers import *
+from gluoncv.utils.parallel import *
+from gluoncv.utils import LRScheduler, LRSequential
+
+from utils import *
+
+
+# Models which were trained with stereo supervision were trained with a nominal
+# baseline of 0.1 units. The KITTI rig has a baseline of 54cm. Therefore,
+# to convert our stereo predictions to real-world scale we multiply our depths by 5.4.
+STEREO_SCALE_FACTOR = 5.4
+
+
+class Trainer:
+    def __init__(self, options, logger):
+        # configuration setting
+        self.opt = options
+        self.logger = logger
+        self.log_path = os.path.join(self.opt.log_dir, self.opt.model_name)
+
+        # checking height and width are multiples of 32
+        assert self.opt.height % 32 == 0, "'height' must be a multiple of 32"
+        assert self.opt.width % 32 == 0, "'width' must be a multiple of 32"
+
+        ################### model initialization ###################
+        self.models = {}
+        self.parameters_to_train = []
+
+        self.num_scales = len(self.opt.scales)
+        self.num_input_frames = len(self.opt.frame_ids)
+        self.num_pose_frames = 2 if self.opt.pose_model_input == "pairs" else self.num_input_frames
+
+        assert self.opt.frame_ids[0] == 0, "frame_ids must start with 0"
+
+        self.use_pose_net = not (self.opt.use_stereo and self.opt.frame_ids == [0])
+
+        if self.opt.use_stereo:
+            self.opt.frame_ids.append("s")
+
+        self.models["encoder"] = monodepthv2.ResnetEncoder(
```

Comment:
Here, we may change the code style a bit. It is better to follow gluoncv style to have `get_model` in train.py. Besides, if you want to resume training or load weights, it is hard to do when you have two separate models.